File: core/src/main/java/javabot/dao/BaseDao.java
package javabot.dao;

import java.util.List;

import org.mongodb.morphia.Datastore;
import org.mongodb.morphia.query.Query;
import com.google.inject.Inject;
import javabot.dao.util.EntityNotFoundException;
import javabot.model.Persistent;
import org.bson.types.ObjectId;

public class BaseDao<T extends Persistent> {
  private final Class<T> entityClass;
  @Inject
  protected Datastore ds;

  protected BaseDao(final Class<T> dataClass) {
    entityClass = dataClass;
  }

  public Query<T> getQuery() {
    return getQuery(entityClass);
  }

  public <U> Query<U> getQuery(Class<U> clazz) {
    return ds.createQuery(clazz);
  }

  public T find(final ObjectId id) {
    return ds.createQuery(entityClass).filter("_id", id).get();
  }

  public List<T> findAll() {
    return ds.createQuery(entityClass).asList();
  }

  private T loadChecked(final ObjectId id) {
    final T persistedObject = find(id);
    if (persistedObject == null) {
      throw new EntityNotFoundException(entityClass, id);
    }
    return persistedObject;
  }

  public void save(final Persistent object) {
    ds.save(object);
  }

  public void delete(final Persistent object) {
    if(object != null) {
      ds.delete(object);
    }
  }

  public void delete(final ObjectId id) {
    delete(loadChecked(id));
  }
}

File: src/main/kotlin/javabot/dao/BaseDao.kt
package javabot.dao

import dev.morphia.Datastore
import dev.morphia.query.FindOptions
import dev.morphia.query.Query
import dev.morphia.query.filters.Filters.eq
import javabot.dao.util.EntityNotFoundException
import javabot.model.Persistent
import org.bson.types.ObjectId
import org.slf4j.LoggerFactory

abstract class BaseDao<T : Persistent>(val ds: Datastore, val entityClass: Class<T>) {
    companion object {
        private val LOG = LoggerFactory.getLogger(BaseDao::class.java)
    }

    fun getQuery(options: FindOptions = FindOptions()): Query<T> = getQuery(entityClass, options)

    fun <U> getQuery(clazz: Class<U>, options: FindOptions = FindOptions()): Query<U> =
        ds.find(clazz, options)

    fun find(id: ObjectId?): T? = ds.find(entityClass).filter(eq("_id", id)).first()

    open fun findAll(): List<T> = ds.find(entityClass).iterator().toList()

    private fun loadChecked(id: ObjectId?) =
        find(id) ?: throw EntityNotFoundException(entityClass, id)

    open fun save(entity: Persistent) {
        ds.save(entity)
    }

    fun delete(entity: Persistent?) {
        entity?.let {
            LOG.debug("deleting entity:  $entity")
            ds.delete(it)
        }
    }

    open fun delete(id: ObjectId?) {
        delete(loadChecked(id))
    }
}
